File: py_src/test/test_sync.py
# -*- coding: utf-8 -*-

from __future__ import print_function
from __future__ import absolute_import

from time import sleep
from sys import version_info

from pytest import (mark, raises)

from .. import (flags, sync)

from .test_mesh import close_all_nodes

if version_info >= (3, ):
    xrange = range


def storage_validation(iters, start_port, num_nodes, encryption, leasing):
    # type: (int, int, int, str, bool) -> None
    for i in xrange(iters):
        print("----------------------Test start----------------------")
        nodes = [
            sync.SyncSocket(
                'localhost',
                start_port + i * num_nodes,
                prot=sync.Protocol('', encryption),
                debug_level=5,
                leasing=leasing)
        ]
        nodes[0].set('store', b"store")
        for j in xrange(1, num_nodes):
            new_node = sync.SyncSocket(
                'localhost',
                start_port + i * num_nodes + j,
                prot=sync.Protocol('', encryption),
                debug_level=5,
                leasing=leasing)
            nodes[-1].connect('localhost', start_port + i * num_nodes + j)
            nodes.append(new_node)
            sleep(0.5)
        print("----------------------Test event----------------------")
        nodes[0]['test'] = b"hello"
        nodes[1][u'测试'] = u'成功'
        nodes[0].update({'array': [1, 2, 3, 4, 5, 6, 7, 8, 9], 'number': 256})
        sleep(num_nodes)
        print("----------------------Test ended----------------------")
        print(nodes[0].id)
        print([len(n.routing_table) for n in nodes])
        for node in nodes[1:]:
            print(node.status, len(node.routing_table))
            assert b"store" == node['store']
            assert b"hello" == node['test']
            assert u'成功' == node[u'测试']
            assert 256 == node['number']
            assert [1, 2, 3, 4, 5, 6, 7, 8, 9] == node['array']
            if leasing:
                with raises(KeyError):
                    node['test'] = b"This shouldn't work"
            with raises(KeyError):
                node['test2']

        close_all_nodes(nodes)


@mark.run(order=4)
def test_storage_leasing_Plaintext(iters=2):
    # type: (int) -> None
    storage_validation(iters, 7100, 3, 'Plaintext', True)


@mark.run(order=4)
def test_storage_leasing_SSL(iters=2):
    # type: (int) -> None
    storage_validation(iters, 7200, 3, 'SSL', True)


@mark.run(order=4)
def test_storage_Plaintext(iters=2):
    # type: (int) -> None
    storage_validation(iters, 7300, 3, 'Plaintext', False)


@mark.run(order=4)
def test_storage_SSL(iters=2):
    # type: (int) -> None
    storage_validation(iters, 7400, 3, 'SSL', False)


def delta_validation(iters, start_port, num_nodes, encryption, leasing):
    # type: (int, int, int, str, bool) -> None
    for i in xrange(iters):
        print("----------------------Test start----------------------")
        nodes = [
            sync.SyncSocket(
                'localhost',
                start_port + i * num_nodes,
                prot=sync.Protocol('', encryption),
                debug_level=5,
                leasing=leasing)
        ]
        for j in xrange(1, num_nodes):
            new_node = sync.SyncSocket(
                'localhost',
                start_port + i * num_nodes + j,
                prot=sync.Protocol('', encryption),
                debug_level=5,
                leasing=leasing)
            nodes[-1].connect('localhost', start_port + i * num_nodes + j)
            nodes.append(new_node)
            sleep(0.5)
        print("----------------------Test event----------------------")
        nodes[0].apply_delta('store', {'seven': 7})
        nodes[1].apply_delta(
            'store', {'array': [1, 2, 3, 4, 5, 6, 7, 8, 9],
                      'number': 256})
        nodes[2].apply_delta('store', {'three': {'three': 'three'}})
        sleep(num_nodes)
        print("----------------------Test ended----------------------")
        print(nodes[0].id)
        print([len(n.routing_table) for n in nodes])
        for node in nodes:
            print(node.status, len(node.routing_table))
            assert node['store'] == {
                'seven': 7,
                'three': {
                    'three': 'three'
                },
                'array': [1, 2, 3, 4, 5, 6, 7, 8, 9],
                'number': 256
            }

        close_all_nodes(nodes)


@mark.run(order=4)
def test_delta_Plaintext(iters=2):
    # type: (int) -> None
    delta_validation(iters, 7500, 3, 'Plaintext', False)


@mark.run(order=4)
def test_delta_SSL(iters=2):
    # type: (int) -> None
    delta_validation(iters, 7600, 3, 'SSL', False)

File: js_src/test/test_sync.js
"use strict";

const assert = require('assert');
const base = require('../base.js');
const sync = require('../sync.js');
var start_port = 44665;

describe('sync', function() {

    describe('SyncSocket', function() {

        let transports = {
            'plaintext': 'Plaintext',
            'SSL/TLS': 'SSL',
            'websocket': 'ws'
        };

        for (let text in transports)    {

            function test_storage(leasing, done)    {
                var node1 = new sync.SyncSocket('localhost', start_port++, new base.Protocol('sync', transports[text]), leasing);
                var node2 = new sync.SyncSocket('localhost', start_port++, new base.Protocol('sync', transports[text]), leasing);

                node1.connect(...node2.addr);

                setTimeout(function()   {
                    node1.set('test', 'value');
                    node2.update({
                        '测试': '成功',
                        'store': 'store',
                        'array': [1,2,3,4,5,6,7,8,9],
                        'number': 256
                    });
                    setTimeout(function()   {
                        assert.ok(node1.get('test'));
                        assert.ok(node2.get('test'));
                        assert.equal(node1.get('测试'), '成功');
                        assert.equal(node2.get('测试'), '成功');
                        assert.equal(node1.get('store'), 'store');
                        assert.equal(node2.get('store'), 'store');
                        assert.equal(node1.get('number'), 256);
                        assert.equal(node2.get('number'), 256);
                        assert.deepEqual(node1.get('array'), [1,2,3,4,5,6,7,8,9]);
                        assert.deepEqual(node2.get('array'), [1,2,3,4,5,6,7,8,9]);
                        done();
                    }, 500);
                }, 250);
            };

            it(`should store values correctly when leasing (over ${text})`, function(done) {
                this.timeout(2500 * (3 && text === 'SSL/TLS' + 1));
                test_storage(true, done);
            });

            it(`should store values correctly when not leasing (over ${text})`, function(done) {
                this.timeout(2500 * (3 && text === 'SSL/TLS' + 1));
                test_storage(false, done);
            });

            function test_delta(done)    {
                var node1 = new sync.SyncSocket('localhost', start_port++, new base.Protocol('sync', transports[text]), false);
                var node2 = new sync.SyncSocket('localhost', start_port++, new base.Protocol('sync', transports[text]), false);
                var node3 = new sync.SyncSocket('localhost', start_port++, new base.Protocol('sync', transports[text]), false);

                node1.connect(...node2.addr);
                node1.connect(...node3.addr);

                setTimeout(function()   {
                    setTimeout(function()   {
                        node1.apply_delta('store', {'seven': 7})
                        node2.apply_delta('store', {'array': [1, 2, 3, 4, 5, 6, 7, 8, 9], 'number': 256})
                        node3.apply_delta('store', {'three': {'three': 'three'}})
                        setTimeout(function()   {
                            let should_be = {'seven': 7,
                                             'array': [1, 2, 3, 4, 5, 6, 7, 8, 9],
                                             'number': 256,
                                             'three': {'three': 'three'}};
                            assert.deepEqual(node1.get('store'), should_be);
                            assert.deepEqual(node2.get('store'), should_be);
                            assert.deepEqual(node3.get('store'), should_be);
                            done();
                        }, 500);
                    }, 250);
                }, 250);
            };

            it(`should apply deltas correctly when not leasing (over ${text})`, function(done) {
                this.timeout(2500 * (3 && text === 'SSL/TLS' + 1));
                test_delta(done);
            });
        }

    });

});
